extensions/amp-link-rewriter/0.1/link-shifter.js

```diff
@@ -0,0 +1,227 @@
+      return;
+    }
+
+    if (this.isInternalLink(htmlElement, trimmedDomain)) {
+      return;
+    }
+
+    this.setTrackingUrl_(htmlElement);
+  }
+
+  /**
+   * Check if the anchor element was already shifted
+   * @param {?Element} htmlElement
+   * @return {boolean}
+   * @private
+   */
+  wasShifted_(htmlElement) {
+    const ctxAttrValue = this.ctxAttrValue_;
+    return (htmlElement.hasAttribute(CTX_ATTR_NAME)) &&
+        (htmlElement.getAttribute(CTX_ATTR_NAME) === ctxAttrValue);
+  }
+
+  /**
+   * Check if the anchor element leads to an internal link
+   * @param {?Element} htmlElement
+   * @param {?string} trimmedDomain
+   * @return {boolean}
+   */
+  isInternalLink(htmlElement, trimmedDomain) {
+    const href = htmlElement.getAttribute('href');
+
+    if (!(href && this.regexDomainUrl_.test(href) &&
+            RegExp.$2 !== trimmedDomain)
+    ) {
+      return true;
+    }
+
+    return false;
+  }
+
+  /**
+   *
+   * @param {?Element} htmlElement
+   */
+  setTrackingUrl_(htmlElement) {
+
+    const oldValHref = htmlElement.getAttribute('href');
+
+    this.viewer_.getReferrerUrl().then(referrerUrl => {
+      const pageAttributes = {
+        referrer: referrerUrl,
+        location: this.viewer_.getResolvedViewerUrl(),
+      };
+
+      htmlElement.href = this.replacePlaceHolders(htmlElement, pageAttributes);
+
+      // If the link has been "activated" via contextmenu,
+      // we have to keep the shifting in mind
+      if (this.event_.type === 'contextmenu') {
+        this.ctxAttrValue_ = CTX_ATTR_VALUE().toString();
+        htmlElement.setAttribute(CTX_ATTR_NAME, this.ctxAttrValue_);
+      }
+
+      this.viewer_.win.setTimeout(() => {
+        htmlElement.href = oldValHref;
+
+        if (htmlElement.hasAttribute(CTX_ATTR_NAME)) {
+          htmlElement.removeAttribute(CTX_ATTR_NAME);
+        }
+
+      }, ((this.event_.type === 'contextmenu') ? 15000 : 500));
+    });
+  }
+
+  /**
+   * @param {?Element} htmlElement
+   * @param {!Object} pageAttributes
+   * @return {string}
+   */
+  replacePlaceHolders(htmlElement, pageAttributes) {
+    const {vars} = this.configOpts_;
+    let {output} = this.configOpts_;
+    const data = {};
+
+    /**
+     * Replace placeholders for anchor attributes
+     * defined in white list constant array
+     */
+    WL_ANCHOR_ATTR.forEach(val => {
+      let attrValue = '';
+
+      if (htmlElement.getAttribute(val)) {
+        attrValue = htmlElement.getAttribute(val);
+      }
+
+      output = output.replace('${' + val + '}', encodeURIComponent(attrValue));
+    });
+
+    /**
+     * Replace placeholders for properties of the document
+     * at the moment only referrer and location
+     */
+    Object.keys(pageAttributes).forEach(key => {
+      let propValue = '';
+
+      if (pageAttributes[key]) {
+        propValue = pageAttributes[key];
+      }
+
+      output = output.replace('${' + key + '}', encodeURIComponent(propValue));
```

Comment:
let's do merge all the vars into one single map and do the replace a single time at the end.